src/extra/datepicker/datepicker-calendar.tsx

```diff
@@ -0,0 +1,306 @@
+import React, { useEffect, useRef, useState } from "react";
+import { useUserLocale } from "../locale";
+import { addMonths, getDate as getDayOfMonth, getDaysInMonth, getWeek, isMonday, set, startOfWeek } from "date-fns";
+import { isEmpty, None, nonEmpty, Option, toOption } from '../../main/option';
+import { adjustDate, DateSettings, getDate, getTimestamp } from "./date-utils";
+import { usePopupPos } from "../../main/popup";
+import { createTimestampState, DatePickerState, CalendarDate } from "./datepicker-exchange";
+
+interface DatepickerCalendarProps {
+  currentState: DatePickerState,
+  currentDateOpt: Option<Date>,
+  dateSettings: DateSettings,
+  setFinalState: (state: DatePickerState) => void,
+  inputRef: React.MutableRefObject<HTMLInputElement | null>,
+  inputBoxRef: React.MutableRefObject<HTMLDivElement | null>
+}
+
+export function DatepickerCalendar({
+  currentState,
+  currentDateOpt,
+  dateSettings,
+  setFinalState,
+  inputRef,
+  inputBoxRef
+}: DatepickerCalendarProps) {
+
+  const popupDate = currentState.popupDate as CalendarDate;
+  const { year, month } = popupDate;
+
+  const pageDate = new Date(year, month);
+
+  const locale = useUserLocale();
+
+  const weeksToShow = 6;
+
+  /*
+   * Popup elements positioning
+  */ 
+  const [popupCalendarRef,setPopupCalendarRef] = useState<HTMLDivElement | null>(null);
+  const [popupCalendarPos] = usePopupPos(popupCalendarRef);
+
+  const [popupMonthRef,setPopupMonthRef] = useState<HTMLDivElement | null>(null);
+  const [popupMonthPos] = usePopupPos(popupMonthRef);
+
+  /*
+   * Months section functionality
+  */
+  const currMonthObj = locale.months.find(monthName => monthName.id === month);
+  const currMonthName = currMonthObj ? currMonthObj.fullName : null;
+
+  function onMonthArrowClick(e: React.MouseEvent<HTMLButtonElement>) {
+    const change = e.currentTarget.dataset.change;
+    if (change) {
+      const newDate = addMonths(new Date(year, month), +change);
+      setFinalState({...currentState, popupDate: getCalendarDate(newDate) });
+    }
+  }
+
+  const [popupMonthShow, setPopupMonthShow] = useState(false);
+
+  const onToggleMonthPopup = () => setPopupMonthShow(prevMonthPopup => !prevMonthPopup);
+
+  const onMonthPopupMiss = () => setPopupMonthShow(false);
+
+  function onMonthChoice(e: React.MouseEvent<HTMLButtonElement>) {
+    const newMonth = e.currentTarget.dataset.month;
+    if (newMonth) setFinalState({ ...currentState, popupDate: { year, month: +newMonth } });
```

Comment:
+?

Reply:
Unary plus is the fastest and preferred way of converting something into a number, because it does not perform any other operations on the number. It can convert string representations of integers and floats, as well as the non-string values true, false, and null. Integers in both decimal and hexadecimal (0x-prefixed) formats are supported. Negative numbers are supported (though not for hex). Using the operator on BigInt values throws a TypeError. If it cannot parse a particular value, it will evaluate to NaN.
(https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Unary_plus)

Поэтому я всегда использую +. Но Number() вроде бы примерно то же самое. Переписал с Number.